### ISSUE TITLE
Missing typeinfos

### ISSUE THREAD

**Issue by disconnect3d** [pwndbg/pwndbg#213]
Some of the binaries might not contain type information about `unsigned long`, `unsigned short` and so call to `gdb.lookup_type(...)` crashes.
I am attaching an example from one of the CTFs.

```
[dc@dc:ransomware_re]$ gdb ./ransomware -q
Loaded 106 commands.  Type pwndbg [filter] for a list.
Reading symbols from ./ransomware...(no debugging symbols found)...done.
pwndbg> run
Starting program: /home/dc/asis/ransomware_re/ransomware 
Exception during func=pwndbg.typeinfo.update ()
Traceback (most recent call last):
  File "/home/dc/installed/pwndbg/pwndbg/events.py", line 122, in caller
    func()
  File "/home/dc/installed/pwndbg/pwndbg/typeinfo.py", line 40, in update
    module.ulong  = gdb.lookup_type('unsigned long')
gdb.error: No type named unsigned long.
Exception during func=pwndbg.events._start_newobjfile (<gdb.NewObjFileEvent object at 0x7f2be28f6468>,)
Traceback (most recent call last):
  File "/home/dc/installed/pwndbg/pwndbg/events.py", line 122, in caller
    func()
  File "/home/dc/installed/pwndbg/pwndbg/events.py", line 178, in _start_newobjfile
    gdb.events.start.on_new_objfile()
  File "/home/dc/installed/pwndbg/pwndbg/events.py", line 60, in on_new_objfile
    function()
  File "/home/dc/installed/pwndbg/pwndbg/events.py", line 128, in caller
    raise e
  File "/home/dc/installed/pwndbg/pwndbg/events.py", line 122, in caller
    func()
  File "/home/dc/installed/pwndbg/pwndbg/typeinfo.py", line 40, in update
    module.ulong  = gdb.lookup_type('unsigned long')
gdb.error: No type named unsigned long.
Python Exception <class 'gdb.error'> No type named unsigned long.: 
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/usr/lib/libthread_db.so.1".
[Inferior 1 (process 26086) exited with code 0377]
pwndbg> 
```

In above example we can actually fetch `size_t` (as we do later `module.size_t = module.ulong` anyway, and `size_t` works with `gdb.lookup_type`). However there are other types that are missing, e.g. `unsigned short`.

Here is the associated code: https://github.com/pwndbg/pwndbg/blob/master/pwndbg/typeinfo.py#L40.
Here is the binary:
[ransomware.zip](https://github.com/pwndbg/pwndbg/files/907687/ransomware.zip)
PS: The name comes from the task name, it won't encrypt your fs. If you want to be fully sure, run it on a VM.

**Comment by zachriggle** [comment]
Unsigned long should work without type information. I'll take a look later.
Thanks!

On Sat, Apr 8, 2017 at 12:35 PM Disconnect3d <notifications@github.com>
wrote:

Some of the binaries might not contain type information about unsigned long
and so gdb.lookup_type('unsigned long') crashes.
I am attaching an example from one of the CTFs.

[dc@dc:ransomware_re]$ gdb ./ransomware -q
Loaded 106 commands.  Type pwndbg [filter] for a list.
Reading symbols from ./ransomware...(no debugging symbols found)...done.
pwndbg> run
Starting program: /home/dc/asis/ransomware_re/ransomware
Exception during func=pwndbg.typeinfo.update ()
Traceback (most recent call last):
  File "/home/dc/installed/pwndbg/pwndbg/events.py", line 122, in caller
    func()
  File "/home/dc/installed/pwndbg/pwndbg/typeinfo.py", line 40, in update
    module.ulong  = gdb.lookup_type('unsigned long')
gdb.error: No type named unsigned long.
Exception during func=pwndbg.events._start_newobjfile
(<gdb.NewObjFileEvent object at 0x7f2be28f6468>,)
Traceback (most recent call last):
  File "/home/dc/installed/pwndbg/pwndbg/events.py", line 122, in caller
    func()
  File "/home/dc/installed/pwndbg/pwndbg/events.py", line 178, in
_start_newobjfile
    gdb.events.start.on_new_objfile()
  File "/home/dc/installed/pwndbg/pwndbg/events.py", line 60, in on_new_objfile
    function()
  File "/home/dc/installed/pwndbg/pwndbg/events.py", line 128, in caller
    raise e
  File "/home/dc/installed/pwndbg/pwndbg/events.py", line 122, in caller
    func()
  File "/home/dc/installed/pwndbg/pwndbg/typeinfo.py", line 40, in update
    module.ulong  = gdb.lookup_type('unsigned long')
gdb.error: No type named unsigned long.
Python Exception <class 'gdb.error'> No type named unsigned long.:
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/usr/lib/libthread_db.so.1".
[Inferior 1 (process 26086) exited with code 0377]
pwndbg>

As we already do:

module.size_t = module.ulong

And there are cases where we can fetch size_t type information (as in this
binary):

pwndbg> py print(gdb.lookup_type('size_t'))
size_t

Maybe we should wrap those module.ulong and module.size_t assignments
<https://github.com/pwndbg/pwndbg/blob/master/pwndbg/typeinfo.py#L40> with
try/catch and try one or the latter?

Here is the binary:
ransomware.zip
<https://github.com/pwndbg/pwndbg/files/907687/ransomware.zip>
PS: The name comes from the task name, it won't encrypt your fs. If you
want to be fully sure, run it on a VM.

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub
<https://github.com/pwndbg/pwndbg/issues/213>, or mute the thread

**Comment by disconnect3d** [comment]
Oh I have edited the post after your reply. There are also other types that are missing:
```
unsigned short
unsigned long long
long long
```

**Comment by disconnect3d** [comment]
@zachriggle It is actually D language binary and it seems they have a little bit different type namings.
The names from here works well:
https://dlang.org/ctod.html#types

I am not sure if we should try C and then D or detect the language/technology the binary was built upon...

**Comment by zachriggle** [comment]
This doesn't really make sense.  Without any binary at all, those types all load fine.

```
pwndbg> file
No executable file now.
No symbol file now.
pwndbg> info file
pwndbg> py print(pwndbg.typeinfo.ulong)
unsigned long
```

**Comment by disconnect3d** [comment]
Merged in #214.
